### PR TITLE
lint: convert format strings linter test to python

### PR DESCRIPTION
**Pull request by Eunoia1729**
Refs #24783

Attempted to keep the style and flow of implementation as it is.
 
### Additional Notes(Optional):
1. There is scope of improvement on how the related files are fetched. In this `git grep` with `subprocess` is still used as I found it to be the simplest. Any pointers on this are appreciated.
2. Removed sort operation on the matching files as I couldn't think of any strong arguments to have it. Any pointers on this are appreciated.
3. Not important, but one small detail is that the previous implementation was storing matched files for all the `function_names` iterated so far. Fixed that in this PR.

### CODE REVIEW

**Comment by laanwj** [comment]
> There is scope of improvement on how the related files are fetched. In this git grep with subprocess is still used as I found it to be the simplest.

I think calling out to `git grep` (and git subcommands in general) is fine. The alternative would be to add a dependency on a git python module. Also, git is a single implementation on every platform (besides version differences).

> Removed sort operation on the matching files as I couldn't think of any strong arguments to have it. Any pointers on this are appreciated.

I don't know the original motivation, the only thing I can think of is that printing errors in with the files in a deterministic sorted order is somewhat-user-friendly. But I don't think it's a big deal.

File: test/lint/lint-format-strings.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        for matching_file in all_matching_files:
+            if not re.search('^src/(leveldb|secp256k1|minisketch|tinyformat|univalue|test/fuzz/strprintf.cpp)', matching_file):
+                matching_files.append(matching_file)
+
+        run_lint_args = ['test/lint/run-lint-format-strings.py',
+                            '--skip-arguments', skip_arguments, function_name,] + matching_files
+        try:
+            subprocess.check_output(run_lint_args, stderr = subprocess.STDOUT)
+        except subprocess.CalledProcessError:
+            exit(1)
```
Comment (laanwj) [inline test/lint/lint-format-strings.py:64]: nit: the convention is to use `sys.exit` instead of the built-in, which is only meant for interactive use.

Comment (Eunoia1729) [reply test/lint/lint-format-strings.py]: Thank you for reviewing !
I've updated the PR according to your suggestions

File: test/lint/lint-format-strings.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2018-2020 The Bitcoin Core developers
```
Comment (vincenzopalazzo) [inline test/lint/lint-format-strings.py:3]:
```suggestion
# Copyright (c) 2018-2022 The Bitcoin Core developers
```

File: test/lint/lint-format-strings.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#
+
+"""
+Lint format strings: This program checks that the number of arguments passed
+to a variadic format string function matches the number of format specifiers
+in the format string.
+"""
```
Comment (vincenzopalazzo) [inline test/lint/lint-format-strings.py:12]:
```suggestion
#!/usr/bin/env python3

"""
Lint format strings: This program checks that the number of arguments passed
to a variadic format string function matches the number of format specifiers
in the format string.

Copyright (c) 2018-2020 The Bitcoin Core developers
Distributed under the MIT software license, see the accompanying
file COPYING or http://www.opensource.org/licenses/mit-license.php.
"""
```

**Review by vincenzopalazzo — COMMENTED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/24802/commits/879c83697d772d31d629b885395b42686e9b1b36

File: test/lint/lint-format-strings.py (hunk 2 above)
Comment (Eunoia1729) [reply test/lint/lint-format-strings.py]: Thank you for the review ! I have updated the PR as per your suggestions.

File: test/lint/lint-format-strings.py (hunk 3 above)
Comment (Eunoia1729) [reply test/lint/lint-format-strings.py]: Previously ported python scripts have the copyright comment in the way existing version of this PR has it. Changing it to the suggested way might look in-consistent with other lint files. Do you still suggest changing ?

**Comment by Eunoia1729** [comment]
Before change:
Screenshot from 2022-04-14 02-36-54

After change:
Screenshot from 2022-04-14 02-54-53

The reason why before version has 3 lines printed and not 2 is due to the issue highlighed in point (3) in the additional notes in the PR header section. IMO the new version is the intended behavior.

**Comment by laanwj** [comment]
Code review ACK 267684ee34d795e4f762c4949fa45d99fe74dde3

File: test/lint/lint-format-strings.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    'vprintf,1',
+    'vsnprintf,1',
+    'vsprintf,1',
+    'WalletLogPrintf,0',
+]
+RUN_LINT_FILE = 'test/lint/run-lint-format-strings.py'
+
+def check_doctest():
+    command = [
+        'python3',
```
Comment (MarcoFalke) [inline test/lint/lint-format-strings.py:39]: wouldn't it be better to use sys.executable?

**Pull request merged**